### PR TITLE
BB2 2241 Update developers.md with new Authorization section content

### PR DESCRIPTION
**Pull request by gmaciolek**
Updated Authorization section with new up-to-date content

**JIRA Ticket:**
[BB2-2241](https://jira.cms.gov/browse/BB2-2241)

**User Story or Bug Summary:**

Updated version of current Authorization section for developers integrating with BB2.0 API.

### What Does This PR Do?

* Updates Authorization section content
* Removes Beneficiary Revoking Access subheading and content under Developer guidelines
* 
### What Should Reviewers Watch For?

If you're reviewing this PR, please check these things, in particular:
* Verify that all code samples in the Authorization section are formatted properly and are accurate

* Links: 
  * check that all links in Authorization section work
  * check that external links open in new tab
  * check that internal links stay in new tab

* Tables in the Authorization section should be formatted properly--like the other tables in the documentation with borders

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following pre-requisite changes have been fully deployed:

### Submitter Checklist

I have gone through and verified that...:

* [x ] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x ] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [ x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [ x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [x ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: _pages/developers.md [hunk 1]
```diff
@@ -128,7 +128,7 @@ You can start using the API right away by following these steps:
 5. Access the synthetic dataset to try additional calls
 
 
-### Step 1: Join the developer sandbox and register a sandbox application
+### <a id="register-application"></a> Step 1: Join the developer sandbox and register a sandbox application
```
Comment (ajshred) [inline _pages/developers.md:131]: I assume this wasn't intended?

File: _pages/developers.md [hunk 2]
```diff
@@ -242,233 +242,230 @@ The first synthetic beneficiary account user is `BBUser00000` with password `PW0
...
-#### Request authorization from user
 
-To allow a user to authorize your application, direct them to the Blue Button 2.0 API authorize endpoint. The request must include the response_type set to code, your application&#39;s client_id, and your application&#39;s redirect_uri. An optional state field that your application can use to identify the authorization request is recommended.
+### Native & mobile applications
 
-~~~
-https://sandbox.bluebutton.cms.gov/v2/o/authorize/?client_id=swBu7LWsCnIRfu530qnfPw1y5vMmER3lAM2L6rq2&amp;redirect_uri=&amp;response_type=code&amp;state=8e896a59f0744a8e93bf2f1f13230be5 
-~~~
-#### Exchange code for token
+The Blue Button API supports the [OAuth 2.0 Authorization Framework](https://tools.ietf.org/html/rfc6749){:target="_blank"} using the [authorization code grant](https://www.rfc-editor.org/rfc/rfc6749#section-1.3.1) with a [confidential client](https://www.rfc-editor.org/rfc/rfc6749#section-2.1){:target="_blank"} type flow. To optimize the security of Medicare enrollees' data during authentication, we do not support the implicit grant or public client types.
```
Comment (ajshred) [inline _pages/developers.md:411]: This external link does not open in a new tab.

**Review by ajshred — CHANGES_REQUESTED** [review]
Just 3 minor items to address.

File: _pages/developers.md (hunk 1 above)
Comment (gmaciolek) [reply _pages/developers.md:131]: It was intended, but I forgot that I did it! I wanted an easy way to link to this section in the beginning of the Authorization section where we say, "To initiate an integration, you'll need the client ID and client secret for your application that were generated when you [registered your application](https://github.com/CMSgov/bluebutton-site-static/blob/7323fc6b1e5c8863f5f7620f14e4056776df980e/_pages/developers.md#register-application)."

**Review by ajshred — APPROVED** [review]
Looks good!  Thanks Gretchen!!!

File: _pages/developers.md [hunk 3]
```diff
@@ -242,233 +242,230 @@ The first synthetic beneficiary account user is `BBUser00000` with password `PW0
...
+* `client_secret`
+* `redirect_uri`
+* `grant_type`: `authorization_code` 
 
-When creating an Application in the sandbox a redirect URI is required. This is the API endpoint on _your_ system that receives the callback from the Blue Button 2.0 API after a beneficiary is passed to the Blue Button 2.0 API to authorize your application.
+##### cURL command
+~~~
+curl -X "https://sandbox.bluebutton.cms.gov/v2/o/token/" \
+-u "swBu7LWsCnIRfu530qnfPw1y5vMmER3lAM2L6rq2:\<client_secret\>" \
+-d "code=TSjqiZCdJwGyytGjz2GzziPfHTJ6z2&grant_type=authorization_code&redirect_uri=http://localhost/testclient/callback](http://localhost/testclient/callback)"
```
Comment (dtisza1) [inline _pages/developers.md:301]:
```suggestion
-d "code=TSjqiZCdJwGyytGjz2GzziPfHTJ6z2&grant_type=authorization_code&redirect_uri=http://localhost:8080/testclient/callback"
```

File: _pages/developers.md [hunk 4]
```diff
@@ -242,233 +242,230 @@ The first synthetic beneficiary account user is `BBUser00000` with password `PW0
...
-```
-URLprotocol://[sub-domain.]domain_name[:port]/path
-```
-### URL Protocol
+To retrieve a new refresh token, POST to the BB2.0 API `/token` endpoint with the following parameters:
 
-Three URL protocols are supported, depending on the purpose:
+* `client_id`
+* `client_secret`
+* `grant_type`: `refresh_token`
```
Comment (dtisza1) [inline _pages/developers.md:323]: I think should add this too?

```suggestion
* `grant_type`: `refresh_token`
* `refresh_token`
```

File: _pages/developers.md [hunk 5]
```diff
@@ -242,233 +242,230 @@ The first synthetic beneficiary account user is `BBUser00000` with password `PW0
...
 
-#### http:// protocol
+##### cURL command
+~~~
+curl -X POST "https://sandbox.bluebutton.cms.gov/v2/o/token/" \
+-u "swBu7LWsCnIRfu530qnfPw1y5vMmER3lAM2L6rq2:\<client_secret\>" \
```
Comment (dtisza1) [inline _pages/developers.md:328]: Backslashes should be removed.

```suggestion
-u "swBu7LWsCnIRfu530qnfPw1y5vMmER3lAM2L6rq2:<client_secret>" \
```

File: _pages/developers.md [hunk 6]
```diff
@@ -242,233 +242,230 @@ The first synthetic beneficiary account user is `BBUser00000` with password `PW0
...
+* `client_id`
+* `client_secret`
+* `redirect_uri`
+* `grant_type`: `authorization_code` 
 
-When creating an Application in the sandbox a redirect URI is required. This is the API endpoint on _your_ system that receives the callback from the Blue Button 2.0 API after a beneficiary is passed to the Blue Button 2.0 API to authorize your application.
+##### cURL command
+~~~
+curl -X "https://sandbox.bluebutton.cms.gov/v2/o/token/" \
+-u "swBu7LWsCnIRfu530qnfPw1y5vMmER3lAM2L6rq2:\<client_secret\>" \
```
Comment (dtisza1) [inline _pages/developers.md:300]: Should remove the backslashes here.

```suggestion
-u "swBu7LWsCnIRfu530qnfPw1y5vMmER3lAM2L6rq2:<client_secret>" \
```

File: _pages/developers.md [hunk 7]
```diff
@@ -242,233 +242,230 @@ The first synthetic beneficiary account user is `BBUser00000` with password `PW0
...
+{
+"access_token": "VD1VaT4IfjXAMlZTS9E4RVXZlkhYG7",
+"expires_in": 36000,
+"token_type": "Bearer",
+"scope": "profile patient/Patient.read patient/Coverage.read patient/ExplanationOfBenefit.read",
+"refresh_token": "7x0VkRQlRU4fRNCQL2vh239nIyucgw",
+"patient": "-20140000000001"
+}
+~~~
+#### Common access token errors
```
Comment (dtisza1) [inline _pages/developers.md:344]: Should change this, since a `refresh` token usage referenced.

```suggestion
#### Common token endpoint errors
```

**Review by dtisza1 — CHANGES_REQUESTED** [review]
@gmaciolek This is looking good.

Had a few change requests.

We should double check the change suggestions after they are rendered.

Sorry that some of the changes got missed in the confluence doc review. 

Thank you!

**Comment by gmaciolek** [comment]
@dtisza1 Thanks for catching all those little details! I had to clean up a bunch of garbage characters going from Confluence to Markdown, so I'm not surprised I missed a few. I committed all your change requests.

**Review by dtisza1 — APPROVED** [review]
@gmaciolek This looks good to me!

**Pull request merged**